### PR TITLE
feat: [CDS-70917]: Create V2 methods in template service to do template resolution on jsonNode instead of yaml

### PR DESCRIPTION
**Pull request by shalini-agr**
## Describe your changes

**Changes:** We have observed that we lose some information when processing on yaml instead of json, so to solve it, we have changed all the processing on yaml to jsonNode in the resolve templates method in service. 

**Testing:** Tested locally by creating templates and using them in pipeline and running them.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48702)

### CODE REVIEW

**Review by archit-harness — COMMENTED** [review]
> Reviewed 2 of 14 files at r1, 3 of 12 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48702)* status: 18 file(s) to review. 0 of 2 approvals obtained. 

___
*[`953-yaml-commons/src/main/java/io/harness/template/yaml/TemplateRefHelper.java` line 43 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mEvJBaKxyX3rrsk1:-NX9mEvJBaKxyX3rrsk2:b-mu4l49) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/953-yaml-commons/src/main/java/io/harness/template/yaml/TemplateRefHelper.java#L43)):*
> ```Java
>   }
> 
>   public boolean hasTemplateRef(JsonNode jsonNode, boolean checkDuplicate) {
> ```

remove this flag



_Code quote:_
```Java
, boolean checkDuplicate
```
___
*[`clients/template-service/src/main/java/io/harness/template/remote/TemplateResourceClient.java` line 115 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mTbYCcroAulXgvR5:-NX9mTbYCcroAulXgvR6:bnfriwi) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/clients/template-service/src/main/java/io/harness/template/remote/TemplateResourceClient.java#L115)):*
> ```Java
>       @Query(value = "AppendInputSetValidator") Boolean appendInputSetValidator);
> 
>   @POST(TEMPLATE_ENDPOINT + "v2/applyTemplates")
> ```

v3 endpoint should be made
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java` line 94 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mfCW4jiuVFpV1xnQ:-NX9mfCW4jiuVFpV1xnR:bli38b5) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java#L94)):*
> ```Java
> 
>   public TemplateMergeResponseDTOWithJsonNode resolveTemplateRefsInPipelineAndAppendInputSetValidators(String accountId,
>       String orgId, String projectId, JsonNode jsonNode, boolean checkForTemplateAccess,
> ```

pipelineJsonNode



_Code quote:_
```Java
jsonNode
```
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java` line 155 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mlKx56zC1qUOBEos:-NX9mlKx56zC1qUOBEot:bau31id) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java#L155)):*
> ```Java
> 
>   private TemplateMergeResponseDTOWithJsonNode resolveTemplateRefsInPipeline(String accountId, String orgId,
>       String projectId, JsonNode jsonNode, boolean checkForTemplateAccess,
> ```

same for this as above
___
*[`template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 522 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9msIsDIW1xv6iGqWB:-NX9msIt5wJK02rTl9_e:b-2xqpu5) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java#L522)):*
> ```Java
> 
>   @POST
>   @Path("v2/applyTemplates")
> ```

v3 endpoint

**Review by archit-harness — COMMENTED** [review]
> Reviewed 4 of 14 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48702)* status: 19 file(s) to review. 0 of 2 approvals obtained. 

___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlUtils.java` line 105 at r3](https://reviewable.io/reviews/harness/harness-core/48702#-NXA90ug7U-XfCFRRDZE:-NXA90ug7U-XfCFRRDZF:bk0rjrb) ([raw file](https://github.com/harness/harness-core/blob/5d4dccc9c751aaec498cb30a504e3fb670d8dced/953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlUtils.java#L105)):*
> ```Java
>   public JsonNode readAsJsonNodeWithCheckDuplicate(String yaml) {
>     try {
>       ObjectMapper mapperWithDuplicate = new ObjectMapper(new YAMLFactory());
> ```

Shouldnt we use same objectMapper of YamlUtils? why we are creating  new one?



_Code quote:_
```Java
ObjectMapper mapperWithDuplicate = new ObjectMapper(new YAMLFactory());
```
___
*[`953-yaml-commons/src/main/java/io/harness/template/yaml/TemplateRefHelper.java` line 43 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mEvJBaKxyX3rrsk1:-NXA7k0v5PPhu8px2pLa:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/953-yaml-commons/src/main/java/io/harness/template/yaml/TemplateRefHelper.java#L43)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

remove this flag
</blockquote></details>

done
___
*[`template-service/service/src/main/java/io/harness/template/helpers/TemplateMergedMaps.java` line 20 at r3](https://reviewable.io/reviews/harness/harness-core/48702#-NXA9qC55uYH2t3ROPiG:-NXA9qC55uYH2t3ROPiH:b-iufgye) ([raw file](https://github.com/harness/harness-core/blob/5d4dccc9c751aaec498cb30a504e3fb670d8dced/template-service/service/src/main/java/io/harness/template/helpers/TemplateMergedMaps.java#L20)):*
> ```Java
> @Builder
> @OwnedBy(HarnessTeam.PIPELINE)
> public class TemplateMergedMaps {
> ```

Always use DTO inside object name, can we rename it better and add comments as well?

**Review by shalini-agr — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48702)* status: 19 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlUtils.java` line 105 at r3](https://reviewable.io/reviews/harness/harness-core/48702#-NXA90ug7U-XfCFRRDZE:-NXATL4U9o-OXpgUAnhT:b-u12hrg) ([raw file](https://github.com/harness/harness-core/blob/5d4dccc9c751aaec498cb30a504e3fb670d8dced/953-yaml-commons/src/main/java/io/harness/pms/yaml/YamlUtils.java#L105)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Shouldnt we use same objectMapper of YamlUtils? why we are creating  new one?
</blockquote></details>

yes, we should use the objectmapper of class here.
___
*[`953-yaml-commons/src/main/java/io/harness/template/yaml/TemplateRefHelper.java` line 43 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mEvJBaKxyX3rrsk1:-NX9x6ESAGsKJKMHnoxQ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/953-yaml-commons/src/main/java/io/harness/template/yaml/TemplateRefHelper.java#L43)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

remove this flag
</blockquote></details>

Done.
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java` line 94 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mfCW4jiuVFpV1xnQ:-NX9xD5_AT20-kSu-OZo:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java#L94)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

pipelineJsonNode
</blockquote></details>

Done.
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java` line 155 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mlKx56zC1qUOBEos:-NX9xHjBAgvsnjb0JOAH:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java#L155)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

same for this as above
</blockquote></details>

Done.
___
*[`clients/template-service/src/main/java/io/harness/template/remote/TemplateResourceClient.java` line 115 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mTbYCcroAulXgvR5:-NX9x6mUAjE0PNhEN8ph:b-webvtw) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/clients/template-service/src/main/java/io/harness/template/remote/TemplateResourceClient.java#L115)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

v3 endpoint should be made
</blockquote></details>

removed new api as discussed
___
*[`template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java` line 522 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9msIsDIW1xv6iGqWB:-NX9xIDkA6aCrivSeac3:b-webvtw) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/template-service/contracts/src/main/java/io/harness/template/resources/NGTemplateResource.java#L522)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

v3 endpoint
</blockquote></details>

removed new api as discussed
___
*[`template-service/service/src/main/java/io/harness/template/helpers/TemplateMergedMaps.java` line 20 at r3](https://reviewable.io/reviews/harness/harness-core/48702#-NXA9qC55uYH2t3ROPiG:-NXATDy34y9yShYGevX6:b-awh4qb) ([raw file](https://github.com/harness/harness-core/blob/5d4dccc9c751aaec498cb30a504e3fb670d8dced/template-service/service/src/main/java/io/harness/template/helpers/TemplateMergedMaps.java#L20)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Always use DTO inside object name, can we rename it better and add comments as well?
</blockquote></details>

removed the class as discussed

**Review by archit-harness — APPROVED** [review]
> Reviewed 2 of 12 files at r2, 6 of 14 files at r3, 8 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48702)* status: 20 file(s) to review. 1 of 2 approvals obtained. 

___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java` line 94 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mfCW4jiuVFpV1xnQ:-NXAW9bc0SMgJ8-mIqoQ:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java#L94)):*
<details><summary><i>Previously, shalini-agr (Shalini Agrawal) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

done
___
*[`pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java` line 155 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mlKx56zC1qUOBEos:-NXAWAbu65XIU6iu46st:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/pipeline-service/service/src/main/java/io/harness/pms/pipeline/service/PMSPipelineTemplateHelper.java#L155)):*
<details><summary><i>Previously, shalini-agr (Shalini Agrawal) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

done
___
*[`clients/template-service/src/main/java/io/harness/template/remote/TemplateResourceClient.java` line 115 at r2](https://reviewable.io/reviews/harness/harness-core/48702#-NX9mTbYCcroAulXgvR5:-NXAWO235JIJ1BU3N-tA:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/eac09db0e85c37ece21559e3f1b67f066ca02781/clients/template-service/src/main/java/io/harness/template/remote/TemplateResourceClient.java#L115)):*
<details><summary><i>Previously, shalini-agr (Shalini Agrawal) wrote…</i></summary><blockquote>

removed new api as discussed
</blockquote></details>

done

**Comment by shalini-agr** [comment]
trigger uts

File: template-service/service/src/main/java/io/harness/template/services/TemplateMergeService.java [hunk 1]
```diff
@@ -20,8 +22,9 @@ String getTemplateInputs(String accountId, String orgIdentifier, String projectI
   TemplateMergeResponseDTO applyTemplatesToYaml(String accountId, String orgId, String projectId, String yaml,
       boolean getMergedYamlWithTemplateField, boolean loadFromCache, boolean appendInputSetValidator);
 
-  TemplateMergeResponseDTO applyTemplatesToYamlV2(String accountId, String orgId, String projectId, String yaml,
-      boolean getMergedYamlWithTemplateField, boolean loadFromCache, boolean appendInputSetValidator);
+  TemplateMergeResponseDTO applyTemplatesToYamlV2(String accountId, String orgId, String projectId,
+      JsonNode pipelineJsonNode, boolean getMergedYamlWithTemplateField, boolean loadFromCache,
```
Comment (utkarshchoubey-harness) [inline template-service/service/src/main/java/io/harness/template/services/TemplateMergeService.java:26]: nit: entityJsonNode would be better name

File: template-service/service/src/main/java/io/harness/template/services/TemplateMergeServiceImpl.java [hunk 2]
```diff
@@ -93,12 +93,13 @@ public TemplateMergeResponseDTO applyTemplatesToYaml(String accountId, String or
...
-      boolean getMergedYamlWithTemplateField, boolean loadFromCache, boolean appendInputSetValidator) {
-    YamlNode yamlNode = TemplateUtils.validateAndGetYamlNode(yaml);
+  public TemplateMergeResponseDTO applyTemplatesToYamlV2(String accountId, String orgId, String projectId,
+      JsonNode pipelineJsonNode, boolean getMergedYamlWithTemplateField, boolean loadFromCache,
+      boolean appendInputSetValidator) {
+    YamlNode pipelineYamlNode = TemplateUtils.validateAndGetYamlNode(pipelineJsonNode);
     TemplateUtils.setupGitParentEntityDetails(accountId, orgId, projectId, null, null);
     Map<String, TemplateEntity> templateCacheMap = new HashMap<>();
-    return getTemplateMergeResponseDTO(accountId, orgId, projectId, yaml, getMergedYamlWithTemplateField, yamlNode,
+    return getTemplateMergeResponseDTO(accountId, orgId, projectId, getMergedYamlWithTemplateField, pipelineYamlNode,
```
Comment (utkarshchoubey-harness) [inline template-service/service/src/main/java/io/harness/template/services/TemplateMergeServiceImpl.java:102]: nit: same name

**Review by utkarshchoubey-harness — APPROVED** [review]

**Comment by shalini-agr** [comment]
trigger uts

**Comment by shalini-agr** [comment]
trigger uts

**Review by utkarshchoubey-harness — APPROVED** [review]

**Review by archit-harness — APPROVED** [review]
> Reviewed 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48702)* status: 21 file(s) to review.

**Comment by shalini-agr** [comment]
trigger review

**Pull request merged**